pkg/ccl/testccl/workload/schemachange/schemachange_test: TestWorkload failed [revert stuck forever]

pkg/ccl/testccl/workload/schemachange/schemachange_test.TestWorkload [failed](https://tanzanite.cluster.engflow.com/invocations/default/cf6d6fb4-151d-497f-a00c-60e9e1157d28?testReportRun=10&testReportShard=1&testReportAttempt=1#targets-Ly9wa2cvY2NsL3Rlc3RjY2wvd29ya2xvYWQvc2NoZW1hY2hhbmdlOnNjaGVtYWNoYW5nZV90ZXN0) on release-24.1 @ [6118b1d327cf70f68056d6f71c08ed0607d433ae](https://github.com/cockroachdb/cockroach/commits/6118b1d327cf70f68056d6f71c08ed0607d433ae):

Fatal error:

```
panic: test timed out after 14m57s
running tests:
	TestWorkload (14m57s)

```
Stack: 

```
goroutine 130561 [running]:
testing.(*M).startAlarm.func1()
	GOROOT/src/testing/testing.go:2366 +0x385
created by time.goFunc
	GOROOT/src/time/sleep.go:177 +0x2d
```
<details><summary>Log preceding fatal error</summary>
<p>

```
   "sql": "CREATE SCHEMA IF NOT EXISTS schema_w0_97 AUTHORIZATION root"
  },
  "COMMIT"
 ],
 "expectedExecErrors": "",
 "expectedCommitErrors": "",
 "message": ""
}
{
 "workerId": 0,
 "clientTimestamp": "05:24:46.903155",
 "ops": [
  "BEGIN",
  {
   "sql": "DROP SEQUENCE IF EXISTS public.seq_w0_96"
  },
  "COMMIT"
 ],
 "expectedExecErrors": "",
 "expectedCommitErrors": "",
 "message": ""
}
{
 "workerId": 0,
 "clientTimestamp": "05:24:47.033492",
 "ops": [
  "BEGIN",
  {
   "sql": "ALTER TABLE schema_w0_73.table_w0_98 ALTER COLUMN \"IrrelevantColumnName\" DROP NOT NULL",
   "expectedExecErr": "42P01"
  }
 ],
 "expectedExecErrors": "42P01",
 "expectedCommitErrors": "",
 "message": "ROLLBACK; Ignoring declarative schema changer not implemented error.: ERROR: *tree.AlterTable not implemented in the new schema changer (SQLSTATE XXUUU)"
}
{
 "workerId": 0,
 "clientTimestamp": "05:24:47.066648",
 "ops": [
  "BEGIN",
  {
   "sql": "DROP INDEX schema_w0_83.table_w0_99@\"IrrelevantIndexName\"",
   "expectedExecErr": "42P01"
  }
 ],
 "expectedExecErrors": "42P01",
 "expectedCommitErrors": "",
 "message": "ROLLBACK; Successfully got expected execution error.: ERROR: relation \"schema_w0_83.table_w0_99\" does not exist (SQLSTATE 42P01)"
}
```
</p>
</details>

Parameters:
 - <code>attempt=1</code>
 - <code>run=10</code>
 - <code>shard=1</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #122785 pkg/ccl/testccl/workload/schemachange/schemachange_test: TestWorkload failed [C-test-failure O-robot P-1 T-sql-foundations branch-release-23.2.5-rc]
- #122745 pkg/ccl/testccl/workload/schemachange/schemachange_test: TestWorkload failed [unexpected value 143] [C-test-failure O-robot P-2 T-sql-foundations branch-release-23.1]
- #121828 pkg/ccl/testccl/workload/schemachange/schemachange_test: TestWorkload failed [error with DROP VALUE followed by DROP SCHEMA in same txn] [C-test-failure O-robot O-schema-testing P-1 T-sql-foundations branch-release-23.2]
</p>
</details>
/cc @cockroachdb/sql-foundations
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestWorkload.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)
</sub>


Jira issue: CRDB-38138

--- a/pkg/sql/type_change.go
+++ b/pkg/sql/type_change.go
@@ -1395,7 +1395,8 @@ func (t *typeChangeResumer) OnFailOrCancel(
 		return nil
 	}(); rollbackErr != nil {
 		switch {
-		case errors.Is(rollbackErr, catalog.ErrDescriptorNotFound):
+		case errors.Is(rollbackErr, catalog.ErrDescriptorNotFound) ||
+			pgerror.GetPGCode(rollbackErr) == pgcode.UndefinedObject:
 			// If the descriptor for the ID can't be found, we assume that another
 			// job executed already and dropped the type.
 			log.Infof(

--- a/pkg/sql/type_change_test.go
+++ b/pkg/sql/type_change_test.go
@@ -14,6 +14,7 @@ import (
 	"context"
 	"fmt"
 	"strings"
+	"sync/atomic"
 	"testing"
 
 	"github.com/cockroachdb/cockroach/pkg/ccl"
@@ -22,6 +23,7 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/sql/pgwire/pgerror"
 	"github.com/cockroachdb/cockroach/pkg/testutils"
 	"github.com/cockroachdb/cockroach/pkg/testutils/serverutils"
+	"github.com/cockroachdb/cockroach/pkg/util/ctxgroup"
 	"github.com/cockroachdb/cockroach/pkg/util/leaktest"
 	"github.com/cockroachdb/cockroach/pkg/util/log"
 	"github.com/cockroachdb/cockroach/pkg/util/syncutil"
@@ -134,6 +136,68 @@ CREATE TYPE d.t AS ENUM();
 			t.Fatalf("expected namespace entries to be cleaned up for type desc %q", name)
 		}
 	}
+}
+
+// TestFailedTypeSchemaChangeIgnoresDrops when a type schema change notices
+// a dropped descriptor during a rollback that is treated as a non-retriable
+// error.
+func TestFailedTypeSchemaChangeIgnoresDrops(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+	defer log.Scope(t).Close(t)
+
+	ctx := context.Background()
+	params, _ := createTestServerParams()
+	startDrop := make(chan struct{})
+	dropFinished := make(chan struct{})
+	cancelled := atomic.Bool{}
+	params.Knobs.SQLTypeSchemaChanger = &sql.TypeSchemaChangerTestingKnobs{
+		RunBeforeExec: func() error {
+			if cancelled.Swap(true) == false {
+				// Kick off a DROP DATABASE job to clean up this descriptor.
+				close(startDrop)
+				<-dropFinished
+				// Fail this schema change so that the rollback logic executes.
+				return jobs.MarkAsPermanentJobError(errors.New("yikes"))
+			} else {
+				return nil
+			}
+		},
+	}
+	// Decrease the adopt loop interval so that retries happen quickly.
+	params.Knobs.JobsTestingKnobs = jobs.NewTestingKnobsWithShortIntervals()
+
+	s, sqlDB, _ := serverutils.StartServer(t, params)
+	defer s.Stopper().Stop(ctx)
+
+	grp := ctxgroup.WithContext(ctx)
+	grp.Go(func() error {
+		defer close(dropFinished)
+		<-startDrop
+		conn, err := sqlDB.Conn(ctx)
+		if err != nil {
+			return err
+		}
+		_, err = conn.ExecContext(ctx, "SET use_declarative_schema_changer = 'off';")
+		if err != nil {
+			return err
+		}
+		_, err = conn.ExecContext(ctx, "DROP DATABASE d CASCADE")
+		return err
+	})
+
+	// Create a type.
+	_, err := sqlDB.Exec(`
+SET use_declarative_schema_changer = 'off';
+CREATE DATABASE d;
+CREATE TYPE d.t AS ENUM('a', 'b', 'c');
+`)
+	require.NoError(t, err)
+
+	// The initial drop should fail.
+	_, err = sqlDB.Exec(`ALTER TYPE d.t DROP VALUE 'c'`)
+	testutils.IsError(err, "yikes")
+
+	require.NoError(t, grp.Wait())
 }
 
 func TestAddDropValuesInTransaction(t *testing.T) {

